specifyweb/frontend/js_src/lib/components/Attachments/PerformAttachmentTask.ts

```diff
@@ -0,0 +1,201 @@
+    readonly onStop: () => void;
+    readonly triggerNow: () => void;
+  }) => JSX.Element | null;
+}): JSX.Element | null {
+  const workRef = React.useRef<AttachmentWorkRef<ACTION>>({
+    mappedFiles: files,
+    uploadPromise: Promise.resolve(0),
+    retrySpec: { 0: 0 },
+  });
+
+  const [workProgress, setWorkProgress] =
+    React.useState<AttachmentWorkProgress>({
+      total: workRef.current.mappedFiles.filter(shouldWork).length,
+      uploaded: 0,
+      type: 'safe',
+      retryingIn: 0,
+    });
+
+  const stop = () => {
+    setWorkProgress((prevState) => ({ ...prevState, type: 'stopping' }));
+  };
+
+  React.useEffect(() => {
+    // Consider the case that React destroys Effect and calls it again. Currently, it only does that
+    // in development to test for function purity. However, still need to guard against that. Solution
+    // is to save promise in a ref and have the next cycle of useEffect wait for it. Additionally, it
+    // should be assumed that files could be uploaded before React gets time to destructor effect completely
+    // since time on main thread could be non-deterministically given.
+
+    if (workProgress.type === 'interrupted') return;
+
+    let destructorCalled = false;
+
+    const setStopped = () =>
+      setWorkProgress((prevState) => ({ ...prevState, type: 'stopped' }));
+
+    const handleProgress = (
+      postUpload: PostWorkUploadSpec<ACTION> | UploadInternalWorkable<ACTION>,
+      currentIndex: number,
+      nextIndex: number
+    ) => {
+      setWorkProgress((progress) => ({
+        ...progress,
+        uploaded: (nextIndex === currentIndex ? 0 : 1) + progress.uploaded,
+      }));
+      workRef.current.mappedFiles = workRef.current.mappedFiles.map(
+        (uploadble, postIndex) =>
+          postIndex === currentIndex ? postUpload : uploadble
+      );
+    };
+
+    const handleStopped = (stoppedIndex: number) => {
+      workRef.current.mappedFiles = workRef.current.mappedFiles.map(
+        (uploadable, index) => ({
+          ...uploadable,
+          status:
+            index >= stoppedIndex && shouldWork(uploadable)
+              ? 'cancelled'
+              : uploadable.status,
+        })
+      );
+      setStopped();
+    };
+    // It may look like this could create a very long promise chain
+    // but the length will always be less than the number of retries
+    // even if the work isn't cancelled.
+    const workPromise = workRef.current.uploadPromise.then((previousIndex) =>
+      previousIndex === undefined
+        ? Promise.resolve(undefined)
+        : new Promise<number | undefined>(async (resolve) => {
+            let nextUploadingIndex = previousIndex;
+            if (workProgress.type === 'stopping') {
+              handleStopped(nextUploadingIndex);
+              resolve(undefined);
+              return;
+            }
+
+            while (nextUploadingIndex < workRef.current.mappedFiles.length) {
```

Comment:
I had a version that use promises instead of a while loop, but couldn't see a good reason to keep that (especially since that'd be more complicated, and harder to debug, than this)

Reply:
sounds good. yes, some code is hard to refactor to pure. all I ask is to prefer functional/pure when possible.

ChatGPT tried it's best, not sure if it's simpler or not:
```tsx
const workPromise = workRef.current.uploadPromise.then((previousIndex) =>
  previousIndex === undefined
    ? Promise.resolve(undefined)
    : new Promise<number | undefined>(async (resolve) => {
        if (workProgress.type === 'stopping') {
          handleStopped(previousIndex);
          resolve(undefined);
          return;
        }

        const result = await workRef.current.mappedFiles.reduce(
          async (accPromise, fileToUpload, currentIndex) => {
            const nextUploadingIndex = await accPromise;
            if (destructorCalled || nextUploadingIndex >= workRef.current.mappedFiles.length) {
              resolve(nextUploadingIndex);
              return nextUploadingIndex;
            }

            if (!shouldWork(fileToUpload)) {
              return nextUploadingIndex + 1;
            }

            const workResult = await workPromiseGenerator(
              fileToUpload,
              () => {
                destructorCalled = true;
                const nextTry = workRef.current.retrySpec[currentIndex] ?? 0;
                workRef.current.retrySpec[currentIndex] = nextTry + 1;
                if (nextTry >= retryTimes.length) {
                  stop();
                  return;
                }
                setWorkProgress((previousProgress) => ({
                  ...previousProgress,
                  type: 'interrupted',
                  retryingIn: retryTimes[nextTry],
                }));
              }
            );

            handleProgress(workResult, currentIndex, nextUploadingIndex + 1);
            return nextUploadingIndex + 1;
          },
          Promise.resolve(previousIndex)
        );

        handleCompletedWork(workRef.current.mappedFiles);
        resolve(undefined);
      })
);
```

pros: no more mutation, no more while loop